pkg/skaffold/runner/dev.go

```diff
@@ -168,7 +168,7 @@ func (r *SkaffoldRunner) Dev(ctx context.Context, out io.Writer, artifacts []*la
 		default:
 			if err := r.monitor.Register(
 				func() ([]string, error) {
-					return build.DependenciesForArtifact(ctx, artifact, r.runCtx, r.artifactStore)
+					return build.DependenciesForArtifactRecursive(ctx, artifacts, artifact, r.runCtx, r.artifactStore)
```

Comment:
Since we're recursing on the `GetDependencies` this operation will get exponentially more expensive with the same artifact dependencies being called multiple times.  Can you try to introduce some sort of cache here, either as a function parameter or converting it to a struct. The cache would have to be flushed on any file change trigger.

Reply:
that's a bit more involving than the trivial fix i made so far, but i can try .. which object should be "owner" of this cache ?
also i don't understand the cache flushing thing, the dependency tree is only built at startup of the "Dev" function aint it, so if i just wipe the cache before entering the for loop on line 157 i should be safe ?

Reply:
i see there is already a GetDependenciesCached seems like i could just use that function (dependencies.go , line 83), no ?

Reply:
> also i don't understand the cache flushing thing, the dependency tree is only built at startup of the "Dev" function aint it, so if i just wipe the cache before entering the for loop on line 157 i should be safe ?

[This](https://github.com/GoogleContainerTools/skaffold/blob/4777aa5e6cd561de8fc7bf50e0abd25d727cc661/pkg/skaffold/runner/dev.go#L172) function gets called everytime the [`Monitor.Run`](https://github.com/GoogleContainerTools/skaffold/blob/63f13d7a7e16506c5247f541fc3c364fda9c4ccb/pkg/skaffold/filemon/monitor.go#L22) method gets called. I think that you need to empty the cache whenever `Monitor.Run` gets triggered otherwise it'll pick up stale dependencies not reflecting the actual files.

Reply:
To test this you should try creating a file in the sample project. With the current implementation of this cache, skaffold probably won't catch this change since the dependency list will be stale and not have the newly created file.